protocol/signalfx/trace_jaeger.go

```diff
@@ -0,0 +1,244 @@
+package signalfx
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/apache/thrift/lib/go/thrift"
+	"github.com/gorilla/mux"
+	jModel "github.com/jaegertracing/jaeger/model"
+	jConverter "github.com/jaegertracing/jaeger/model/converter/thrift/jaeger"
+	jThrift "github.com/jaegertracing/jaeger/thrift-gen/jaeger"
+	"github.com/opentracing/opentracing-go/ext"
+	zModel "github.com/openzipkin/zipkin-go/model"
+	"github.com/signalfx/golib/log"
+	"github.com/signalfx/golib/pointer"
+	"github.com/signalfx/golib/sfxclient"
+	"github.com/signalfx/golib/trace"
+	"github.com/signalfx/golib/web"
+)
+
+// JaegerThriftTraceDecoderV1 decodes Jaeger thrift spans to structs
+type JaegerThriftTraceDecoderV1 struct {
+	Logger log.Logger
+	Sink   trace.Sink
+}
+
+var errInvalidJaegerTraceFormat = errors.New("invalid Jaeger format")
+
+const (
+	// JaegerV1 binary thrift protocol
+	JaegerV1 = "jaeger_thrift_v1"
+)
+
+func setupThriftTraceV1(ctx context.Context, r *mux.Router, sink Sink, logger log.Logger, httpChain web.NextConstructor) sfxclient.Collector {
+	handler, st := setupChain(ctx, sink, JaegerV1, func(s Sink) ErrorReader {
+		return &JaegerThriftTraceDecoderV1{Logger: logger, Sink: sink}
+	}, httpChain, logger)
+
+	r.Path(DefaultTracePathV1).Methods("POST").Headers("Content-Type", "application/x-thrift").Handler(handler)
+	r.Path(DefaultTracePathV1).Methods("POST").Headers("Content-Type", "application/vnd.apache.thrift.binary").Handler(handler)
+
+	return st
+}
+
+// Code inspired by
+// https://github.com/jaegertracing/jaeger/blob/89f3ccaef21d256728f02ec9d73b31f9c3bde71a/cmd/collector/app/http_handler.go#L61
+func (decoder *JaegerThriftTraceDecoderV1) Read(ctx context.Context, req *http.Request) error {
+	bodyBytes, err := ioutil.ReadAll(req.Body)
+	_ = req.Body.Close()
+	if err != nil {
+		return errors.New("could not read request body")
+	}
+
+	tDeserializer := thrift.NewTDeserializer()
+	batch := &jThrift.Batch{}
+	if err = tDeserializer.Read(batch, bodyBytes); err != nil {
+		return errInvalidJaegerTraceFormat
+	}
+
+	spans := convertJaegerBatch(batch)
+
+	return decoder.Sink.AddSpans(ctx, spans)
+}
+
+func convertJaegerBatch(batch *jThrift.Batch) []*trace.Span {
+	spans := make([]*trace.Span, len(batch.Spans), len(batch.Spans))
```

Comment:
don't need both len and cap if they're the same, just make([]*trace.Span, len(batch.Spans)) will set them both to the same value.